KeeTrayTOTP/Menu/MenuItemProvider.cs

```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Windows.Forms;
+using KeePass.Plugins;
+
+namespace KeeTrayTOTP.Menu
```

Comment:
* It seems no one derives MenuItem provider, or is that a plan for the future? Otherwise you can seal it.
* You can specify the exact types for  TrayMenuItemProvider, EntryMenuItemProvider, MainMenuItemProvider instead of using MenuItemProviderBase.
* The constructor always sets TrayMenuItemProvider, EntryMenuItemProvider, MainMenuItemProvider, so GetMenuItem can immediately call `provider.ProvideMenuItem()` without the null check. The protected members can then go as well.

```csharp
using System;
using System.Windows.Forms;
using KeePass.Plugins;

namespace KeeTrayTOTP.Menu
{
    public sealed class MenuItemProvider : IDisposable
    {
        private bool _isDisposed;
        internal TrayMenuItemProvider TrayMenuItemProvider { get; private set; }
        internal EntryMenuItemProvider EntryMenuItemProvider { get; private set; }
        internal MainMenuItemProvider MainMenuItemProvider { get; private set; }

        public MenuItemProvider(KeeTrayTOTPExt plugin, IPluginHost pluginHost)
        {
            if (pluginHost.CustomConfig.GetBool(KeeTrayTOTPExt.setname_bool_LegacyTrayMenuProvider_Enable, false))
            {
                TrayMenuItemProvider = new LegacyTrayMenuItemProvider(plugin, pluginHost);
            }
            else
            {
                TrayMenuItemProvider = new TrayMenuItemProvider(plugin, pluginHost);
            }

            EntryMenuItemProvider = new EntryMenuItemProvider(plugin, pluginHost);
            MainMenuItemProvider = new MainMenuItemProvider(plugin);
        }

        internal ToolStripMenuItem GetMenuItem(PluginMenuType type)
        {
            switch (type)
            {
                // Provide a menu item for the main location(s)
                case PluginMenuType.Main:
                    return MainMenuItemProvider.ProvideMenuItem();

                case PluginMenuType.Entry:
                    return EntryMenuItemProvider.ProvideMenuItem();

                case PluginMenuType.Tray:
                    return TrayMenuItemProvider.ProvideMenuItem();

                default:
                    return null; // No menu items in other locations
            }
        }

        public void Dispose()
        {
            // Dispose of unmanaged resources.
            Dispose(true);
            // Suppress finalization.
            GC.SuppressFinalize(this);
        }

        public void Dispose(bool disposing)
        {
            if (_isDisposed) return;

            if (disposing)
            {
                TrayMenuItemProvider.Terminate();
                EntryMenuItemProvider.Terminate();
                MainMenuItemProvider.Terminate();
            }

            _isDisposed = true;
        }
    }
}
```

Reply:
Yes I though maybe someone wants to add a new plugin, just to change the behavior of the menu items. But after a second thought, this is pretty unlikley